tmt/trying.py

```diff
@@ -0,0 +1,316 @@
+""" Easily try tests and experiment with guests """
+
+import textwrap
+from typing import Any
+
+import click
+import fmf
+from fmf.utils import listed
+
+import tmt
+import tmt.log
+import tmt.steps
+import tmt.steps.provision
+import tmt.utils
+from tmt import Plan
+from tmt.templates import DEFAULT_PLAN, DEFAULT_PLAN_NAME
+from tmt.utils import MetadataError, Path
+
+USER_PLAN_NAME = "/user/plan"
+
+
+class Try(tmt.utils.Common):
+
+    def __init__(
+            self,
+            *,
+            tree: tmt.Tree,
+            logger: tmt.log.Logger,
+            **kwargs: Any) -> None:
+        """ Just store the tree """
+
+        super().__init__(logger=logger, **kwargs)
+
+        self.tree = tree
+        self.tests: list[tmt.Test] = []
+        self.plans: list[Plan] = []
+        self.image_and_how = self.opt("image_and_how")
+
+    def check_tree(self) -> None:
+        """ Make sure there is a sane metadata tree """
+
+        # Both tree and root should be defined
+        try:
+            if self.tree and self.tree.root:
+                return
+
+        # Create a dumb fmf Tree if no metadata around
+        except MetadataError:
+            self.tree.tree = fmf.Tree({"nothing": "here"})
+
+    def check_tests(self) -> None:
+        """ Check for available tests """
+
+        # Search for tests according to provided names
+        test_names = list(self.opt("test"))
+        if test_names:
+            self.tests = self.tree.tests(names=test_names)
+            if not self.tests:
+                raise tmt.utils.GeneralError(
+                    f"No test matching '{listed(test_names)}' found.")
+
+        # Default to tests under the current working directory
+        else:
+            if not self.tree.root:
+                self.debug("No fmf tree root, no tests.")
+                return
+            relative_path = Path(".").relative_to(self.tree.root)
+            test_names = [f"^/{relative_path}"]
+            self.tests = self.tree.tests(names=test_names)
+            if not self.tests:
+                self.warn(f"No tests found under the '{relative_path}' directory.")
+
+        # Short debug info about what was found
+        self.debug("Test name filter", listed(test_names, quote="'"))
+        self.debug("Matching tests found\n" + tmt.utils.format_value(self.tests))
+
+        # Inject the test filtering options into the Test class
+        options = {"names": [f"^{test.name}$" for test in self.tests]}
+        tmt.Test.store_cli_invocation(context=None, options=options)
+
+    def get_default_plans(self, run: tmt.base.Run) -> list[Plan]:
+        """ Get default plan from user config or the standard template """
+
+        # Check user config for custom default plans. Search for all
+        # plans starting with the default user plan name (there might be
+        # more than just one).
+        try:
+            config_tree = tmt.utils.Config().fmf_tree
+            plan_name = USER_PLAN_NAME
```

Comment:
Is `USER_PLAN_NAME` a name or a pattern? If it's a name, then it should be wrapped by `re.escape()` when injected into various patterns below, to be matched literaly. If the name ever changes to something with riskier characters, without escaping them and making them literals we might observe some unexpected results.

Reply:
Escaped. Fixed in 842df674.